### PR TITLE
Add mixin for longer chat messages

### PR DESCRIPTION
**Pull request by serenibyss**
Adds a two class mixin to allow raising the chat message length limit up to (currently) 512. Newer versions of Minecraft raised this limit to 256, so that is what I chose as the default value. The 512 limit is arbitrary, and I'm not sure if there are larger implications of it being this high, or if it could be higher.

I'm not the most familiar with mixin, so please let me know if there is any problems with remapping especially since I used a deobfuscated name in two of the mixin targets.

### CODE REVIEW

**Comment by boubou19** [comment]
what happens if server and client have different values?

**Comment by serenibyss** [comment]
> what happens if server and client have different values? Or if 2 clients haven't the same value?

This is a problem, as if the server's value is smaller than the client, and the client sends a message that's too long, it will kick the client from the server. I think there's 2 different approaches to handling it, either just force the chat length to 256 like newer Minecraft versions (or some other agreed upon number), or somehow synchronize the server config value to the client on join (not sure how easy this is)

**Comment by boubou19** [comment]
imo best would be for clients to be able to automatically break down chat messages into multiple ones if they have a value below the server one

**Comment by Alexdoru** [comment]
> what happens if server and client have different values? Or if 2 clients haven't the same value?

there is no problem if two clients have different values, but it will shorten the message to the server value if the server and client have different values

**Comment by Alexdoru** [comment]
> > what happens if server and client have different values? Or if 2 clients haven't the same value?
> 
> This is a problem, as if the server's value is smaller than the client, and the client sends a message that's too long, it will kick the client from the server. I think there's 2 different approaches to handling it, either just force the chat length to 256 like newer Minecraft versions (or some other agreed upon number), or somehow synchronize the server config value to the client on join (not sure how easy this is)

imo should just remove the config option to choose the lenght and hardcode it to 256

**Comment by Alexdoru** [comment]
> Adds a two class mixin to allow raising the chat message length limit up to (currently) 512. Newer versions of Minecraft raised this limit to 256, so that is what I chose as the default value. The 512 limit is arbitrary, and I'm not sure if there are larger implications of it being this high, or if it could be higher.
> 
> I'm not the most familiar with mixin, so please let me know if there is any problems with remapping especially since I used a deobfuscated name in two of the mixin targets.

I pushed two commits to correct your pr

File: src/main/java/com/mitchej123/hodgepodge/LoadingConfig.java [hunk 1]
```diff
@@ -200,6 +201,7 @@ public LoadingConfig(File file) {
         arabicNumbersForEnchantsPotions = config.get(Category.TWEAKS.toString(), "arabicNumbersForEnchantsPotions", false, "Uses arabic numbers for enchantment levels and potion amplifier levels instead of roman numbers").getBoolean();
         bedMessageAboveHotbar = config.get(Category.TWEAKS.toString(), "bedMessageAboveHotbar", true, "Show \"cannot sleep\" messages above hotbar").getBoolean();
         chatLength = Math.max(100, Math.min(32767, config.get(Category.TWEAKS.toString(), "chatLength", 8191, "Amount of chat lines kept [100(Vanilla) - 32767]").getInt()));
+        messageLength = Math.max(100, Math.min(512, config.get(Category.TWEAKS.toString(), "messageLength", 256, "Length of messages sent in chat [100(Vanilla) - 512").getInt()));
```
Comment (Alexdoru) [inline src/main/java/com/mitchej123/hodgepodge/LoadingConfig.java:204]: can you remove this setting and just hardcode it to 256 please

**Review by Alexdoru — CHANGES_REQUESTED** [review]

File: src/main/java/com/mitchej123/hodgepodge/LoadingConfig.java (hunk 1 above)
Comment (glowredman) [reply src/main/java/com/mitchej123/hodgepodge/LoadingConfig.java:204]: Or sync the value from server to client

Comment (Alexdoru) [reply src/main/java/com/mitchej123/hodgepodge/LoadingConfig.java:204]: > Or sync the value from server to client

that's too much effort for such a feature

**Comment by Alexdoru** [comment]
I'm ok with this now if anyone else wants to look at it

**Comment by Dream-Master** [comment]
@serenibyss

**Comment by serenibyss** [comment]
If the server has the setting off, and a client joins with the setting on, it would still disconnect the player if an oversized message is sent

**Comment by Dream-Master** [comment]
@serenibyss are there any way to fix it?

**Comment by Dream-Master** [comment]
@serenibyss can you resolve the conflict?

File: src/main/java/com/mitchej123/hodgepodge/mixins/Mixins.java [hunk 2]
```diff
@@ -83,6 +83,12 @@ public enum Mixins {
     TRANSPARENT_CHAT(new Builder("Transparent Chat").setPhase(Phase.EARLY)
             .addMixinClasses("minecraft.MixinGuiNewChat_TransparentChat").setSide(Side.CLIENT)
             .setApplyIf(() -> Common.config.transparentChat).addTargetedMod(TargetedMod.VANILLA)),
+    LONGER_MESSAGES_CLIENT(new Builder("Longer Messages").setPhase(Phase.EARLY)
+            .addMixinClasses("minecraft.MixinGuiChat_LongerMessages").setApplyIf(() -> Common.config.longerSentMessages)
+            .addTargetedMod(TargetedMod.VANILLA).setSide(Side.CLIENT)),
+    LONGER_MESSAGES_SERVER(new Builder("Longer Messages Server Side").setPhase(Phase.EARLY)
+            .addMixinClasses("minecraft.MixinC01PacketChatMessage_LongerMessages").setApplyIf(() -> true)
```
Comment (Alexdoru) [inline src/main/java/com/mitchej123/hodgepodge/mixins/Mixins.java:90]: I think the reason we put config options everywhere is so we can have maximum compatibility with other mods and disable features if they cause crashes for certain installations. The proper way to code that feature would be to make the server send a packet to the client informing it of the chat message length authorized on that server. But ngl I don't know how to do that.

**Review by Alexdoru — CHANGES_REQUESTED** [review]

File: src/main/java/com/mitchej123/hodgepodge/mixins/Mixins.java (hunk 2 above)
Comment (Caedis) [reply src/main/java/com/mitchej123/hodgepodge/mixins/Mixins.java:90]: Same. This is what I could come up with.

**Comment by Caedis** [comment]
I believe that is the proper way to add config syncing.
Tested locally and watching it set the config to false from the server on the client when it was true on the client.
That then prevented sending a message over 100 characters.

This should also allow for future (or current) config options to be synced.

**Comment by Alexdoru** [comment]
Since I never worked with custom packets, I can't review that PR sorry, but I am glad you could figure it out :p

File: src/main/java/com/mitchej123/hodgepodge/net/HandlerConfigSync.java [hunk 3]
```diff
@@ -0,0 +1,17 @@
+package com.mitchej123.hodgepodge.net;
+
+import com.mitchej123.hodgepodge.Common;
+
+import cpw.mods.fml.common.network.simpleimpl.IMessage;
+import cpw.mods.fml.common.network.simpleimpl.IMessageHandler;
+import cpw.mods.fml.common.network.simpleimpl.MessageContext;
+
+public class HandlerConfigSync implements IMessageHandler<MessageConfigSync, IMessage> {
```
Comment (miozune) [inline src/main/java/com/mitchej123/hodgepodge/net/HandlerConfigSync.java:9]: Optional: many mods implement `IMessage` and `IMessageHandler` with one class

File: src/main/java/com/mitchej123/hodgepodge/mixins/Mixins.java [hunk 4]
```diff
@@ -84,6 +84,14 @@ public enum Mixins {
     TRANSPARENT_CHAT(new Builder("Transparent Chat").setPhase(Phase.EARLY)
             .addMixinClasses("minecraft.MixinGuiNewChat_TransparentChat").setSide(Side.CLIENT)
             .setApplyIf(() -> Common.config.transparentChat).addTargetedMod(TargetedMod.VANILLA)),
+    // config handled in mixin due to server->client config sync
+    LONGER_MESSAGES_CLIENT(new Builder("Longer Messages").setPhase(Phase.EARLY)
```
Comment (miozune) [inline src/main/java/com/mitchej123/hodgepodge/mixins/Mixins.java:88]: `Longer Messages Client Side`?

File: src/main/java/com/mitchej123/hodgepodge/net/HandlerConfigSync.java (hunk 3 above)
Comment (Caedis) [reply src/main/java/com/mitchej123/hodgepodge/net/HandlerConfigSync.java:9]: Can you show me an example?

Comment (Caedis) [reply src/main/java/com/mitchej123/hodgepodge/net/HandlerConfigSync.java]: Found one and implemented

**Review by miozune — APPROVED** [review]

**Pull request merged**